test/gtest/ucs/test_hash_perf.cc

```diff
@@ -0,0 +1,281 @@
+    }
+
+    void init_storage() {
+        obj = (SGLIB_hash_type **) ucs_malloc(sizeof(*obj) * SGLIB_HASH_SIZE, "sglib_hash");
+        sglib_hashed_SGLIB_hash_type_init(obj);
+    }
+
+    void add_elem(const uint64_t key) {
+        SGLIB_hash_type *ep = create_element<SGLIB_hash_type>(key);
+        sglib_hashed_SGLIB_hash_type_add(obj, ep);
+    }
+
+    Hash_type *find_elem(const uint64_t key) const {
+        SGLIB_hash_type ep, *ep_found = 0;
+        ep.dest_uuid = key;
+        ep_found = sglib_hashed_SGLIB_hash_type_find_member(obj, &ep);
+        EXPECT_TRUE(ep_found);
+        return ep_found;
+    }
+
+    void del_elems() {
+        struct sglib_hashed_SGLIB_hash_type_iterator it;
+        for (SGLIB_hash_type *local_it = sglib_hashed_SGLIB_hash_type_it_init(&it, obj); local_it != NULL; local_it = sglib_hashed_SGLIB_hash_type_it_next(&it)) {
+            sglib_hashed_SGLIB_hash_type_delete(obj, local_it);
+            free(local_it);
+        }
+        free(obj);
+    }
+
+    SGLIB_hash_type **obj;
+};
+
+class test_hash_perf : public ucs::test {
+};
+
+UCS_TEST_F(test_hash_perf, perf_compare) {
+
+    size_t trip_counts[] = {1, 2, 8, 128, 1024, 32768, 262144, 1048576, 0};
+    ucs_time_t khash_lookup_ns_1024 = 0;
+    ucs_time_t sglib_lookup_ns_1024 = 0;
+    Perf_compare_base *perf_compare_khash_ptr = new Perf_compare_khash;
+    Perf_compare_base *perf_compare_sglib_ptr = new Perf_compare_sglib;
+    Perf_compare_base *hashes[] = {
+        perf_compare_khash_ptr,
+        perf_compare_sglib_ptr,
+        new Perf_compare_map,
+        new Perf_compare_unordered_map,
+        NULL
+    };
+
+    if (ucs::test_time_multiplier() > 1) {
+        UCS_TEST_SKIP;
+    }
+
+    generate_keys();
+
+    UCS_TEST_MESSAGE << ":    elements   :init  :lookup:remove";
+    for (int i = 0; hashes[i] != NULL; ++i) {
+        Perf_compare_base *cur_hash = hashes[i];
+        for (int j = 0; trip_counts[j] > 0; ++j) {
+            size_t num_elems = trip_counts[j];
+
+            ucs_time_t insert_ns = cur_hash->initialize(num_elems);
+            ucs_time_t lookup_ns = cur_hash->lookup(num_elems);
+            ucs_time_t remove_ns = cur_hash->cleanup(num_elems);
+
+            UCS_TEST_MESSAGE << FIELD_ALIGN(6) << cur_hash->get_name()
+                             << FIELD_ALIGN(8) << num_elems
+                             << FIELD_ALIGN(6) << insert_ns
+                             << FIELD_ALIGN(6) << lookup_ns
+                             << FIELD_ALIGN(6) << remove_ns;
+            if (1024 == num_elems) {
+                if (cur_hash == perf_compare_khash_ptr ) {
+                    khash_lookup_ns_1024 = lookup_ns;
+                }
+                if (cur_hash == perf_compare_sglib_ptr ) {
+                    sglib_lookup_ns_1024 = lookup_ns;
+                }
+            }
+        }
+        delete cur_hash;
+    }
+    EXPECT_LT(khash_lookup_ns_1024, 400.0);
+    EXPECT_LT(sglib_lookup_ns_1024, 400.0);
```

Comment:
Perhaps we need to do more performance check points here. For example, we can compare std::map and khash performance at some point.